File: cpp/leetcode/130.cpp
#include <iostream>

using namespace std;
class Solution {
public:
    void solve(vector<vector<char>>& board) {
        if(board.empty() || board[0].empty()) return ;
        int m = board.size(), n = board[0].size();
        for(int i = 0; i < m; i++){
            if(board[i][0] == 'O') dfs(i, 0, board);
            if(board[i][n-1] == 'O') dfs(i, n-1, board);
        }
        for(int i = 0; i < n; i++){
            if(board[0][i] == 'O') dfs(0, i, board);
            if(board[m-1][i] == 'O') dfs(m-1, i, board);
        }
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                if(board[i][j] == 'O') board[i][j] = 'X';
                else if(board[i][j] == '#') board[i][j] = 'O';
            }
        }
    }

    void dfs(int i, int j, vector<vector<char>> & board){
        int m = board.size(), n = board[0].size();
        if(board[i][j] == 'O') board[i][j] = '#';
        if(i > 1 && board[i-1][j] == 'O') dfs(i-1, j, board); // i > 0
        if(i < m-1 && board[i+1][j] == 'O') dfs(i+1, j, board);
        if(j > 1 && board[i][j-1] == 'O') dfs(i, j-1, board); // j > 0
        if(j < n-1 && board[i][j+1] == 'O') dfs(i, j+1, board);
    }
};

File: python/leetcode/130.py
class Solution:
    def solve(self, board):
        if not board: return
        m, n = len(board), len(board[0])
        for j in range(n):
            self.fill(board, 0, j)
            self.fill(board, m - 1, j)
        for i in range(m):
            self.fill(board, i, 0)
            self.fill(board, i, n - 1)
        for i in range(m):
            for j in range(n):
                if board[i][j] == 'O': board[i][j] = 'X'
                elif board[i][j] == '#': board[i][j] = 'O'

    def fill(self, board, i, j):
        m, n = len(board), len(board[0])
        queue = [(i, j)]
        while queue:
            row, col = queue.pop(0)
            if row < 0 or row >= m or col < 0 or col >= n or board[row][col] != 'O':
                continue
            board[row][col] = '#'
            queue.append((row - 1, col))
            queue.append((row + 1, col))
            queue.append((row, col - 1))
            queue.append((row, col + 1))